sw/device/tests/sram_ctrl_main_scrambled_access_test.c

```diff
@@ -0,0 +1,177 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/base/multibits.h"
+#include "sw/device/lib/base/stdasm.h"
+#include "sw/device/lib/dif/dif_rstmgr.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/rstmgr_testutils.h"
+#include "sw/device/lib/testing/sram_ctrl_testutils.h"
+#include "sw/device/lib/testing/test_framework/test_main.h"
+#include "sw/device/lib/testing/test_framework/test_status.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "rstmgr_regs.h"     // Generated.
+#include "sram_ctrl_regs.h"  // Generated.
+
+const test_config_t kTestConfig;
+
+static dif_sram_ctrl_t sram_ctrl;
+
+/**
+ * MAIN SRAM scrambling test buffer.
+ *
+ * This buffer is initialized with a known pattern, which will be overwritten
+ * with a pseudo-random data as the result of MAIN SRAM scrambling.
+ */
+static volatile uint32_t
+    sram_ctrl_main_scramble_buffer[SRAM_CTRL_TESTUTILS_DATA_NUM_WORDS];
+
+/**
+ * Test pattern to be written and read from SRAM.
+ */
+static const sram_ctrl_testutils_data_t kRamTestPattern1 = {
+    .words =
+        {
+            0xA5A5A5A5,
+            0xA5A5A5A5,
+            0xA5A5A5A5,
+            0xA5A5A5A5,
+        },
+};
+
+/**
+ * Test pattern to be written and read from SRAM.
+ */
+static const sram_ctrl_testutils_data_t kRamTestPattern2 = {
+    .words =
+        {
+            0x5A5A5A5A,
+            0x5A5A5A5A,
+            0x5A5A5A5A,
+            0x5A5A5A5A,
+        },
+};
+
+/**
+ * Performs scrambling, saves the test relevant data and resets the system.
+ *
+ * MAIN SRAM will be overwritten with the pseudo-random data (including .bss
+ * and .data segments), which means that the code must be written in assembly.
+ *
+ * This function saves contents of the `sram_ctrl_main_scramble_buffer` in the
+ * RETENTION SRAM, which maintains the power across the system reboot.
+ */
+static noreturn void main_sram_scramble(void) {
+  asm volatile(
+      // Request the new scrambling key for MAIN SRAM.
+      "li %[kGeneralRW], 0x1;                                              \n"
+      "sw %[kGeneralRW], %[kSramCtrlOffset](%[kSramCtrlRegsBase]);         \n"
+
+      // Busy loop - waiting for scrambling to finish.
+      ".L_SCRAMBLING_BUSY_LOOP:;                                           \n"
```

Comment:
nit: lowercase for labels (i.e. `.L_scrambling_busy_loop`). Not important but just makes them visually a bit different to preprocessor constants.